specification/containerservice/resource-manager/Microsoft.ContainerService/aks/preview/2023-08-02-preview/managedClusters.json

```diff
@@ -3943,6 +4022,10 @@
           {
             "value": "AvailabilitySet",
             "description": "Use of this is strongly discouraged."
+          },
+          {
```

Comment:
I feel like we're growing in options here too quickly. We are having `Flex` and `VirtualMachines` and they both seem very similar.

I like the `Flex` name, but wondering if we can collapse `Flex` + `VirtualMachines` into a single `Type`? It seems strange to have them separate given how close they are in terms of impl and capabilities? Part of the whole idea/win of `VirtualMachines` was to enable this `Flex`-like behavior.

Reply:
For example, we could just make `VirtualMachines` become `Flex` and use this new specification mode, and not have `VirtualMachines` at all. That also would really help users understand the value-add of `VirtualMachines` (now `Flex`) because it's clearly paving the way to mixed-sku pools.

Reply:
We want to drop virtualMachines in naming as cx shouldn't care about the what's the format of the actual machine. VMAS/VMSS/VM is aks implementation detail

We want to preview both - flex will come with an experimental non blocking scaling feature.
We'll try to merge before GA.

Reply:
> We'll try to merge before GA.

For me at least, this feels like a must. I'm not really sold on both of them even in preview, but maybe it's OK if it's temporary. Half of the point of `VirtualMachines` mode was to pave the way for mixed-sku pools. If it's _never_ going to have them because they're going into `Flex` then why would you ever want to use `VirtualMachines` mode over `Flex`? Flex is just a superset of `VirtualMachines` at that point, isn't it?

cc @zqingqing1 

See also https://github.com/Azure/azure-rest-api-specs/pull/26034 as well.

Wondering if it makes sense when `Flex` launches to just remove `VirtualMachines` from preview, which is a breaking change but OK since preview?

Reply:
Yes the goal is to merge before GA instead of Preview - as the two types come with different feature sets and scaling pattern. So separate the preview progress.

Reply:
> Wondering if it makes sense when Flex launches to just remove VirtualMachines from preview, which is a breaking change but OK since preview?

I don't agree to drop virtualmachines unless it's been through the deprecation process.

Reply:
There is no deprecation process for preview features @weinong, see aka.ms/bcforapi, section on previews:
> However, a new preview or GA version can make breaking changes as compared to the previous preview version. Once a new preview or GA version is available, notify customers that the previous preview version will stop working entirely in 
90 days. You don’t need to go through the Breaking Change or Retirement process.

As long as the existing API is supported for 90d (we do a year), we are free to remove preview features.

Reply:
We also don't have to drop those VM APs  - we can just migrate them to flex so with later API version they'll just see those as flex aps.

Reply:
+1, didn't mean to imply we should just magically disappear those APs or anything, just that we should come up with some way (during preview) to flow VM pools into Flex (if Flex is really the future).

I don't care that much if `VirtualMachines` or `Flex` wins (I like the Flex name more), I just think going from 2 types (VMAS, VMSS) where 1 is super discouraged to 4 types where 3 are "OK" is going to produce a lot of customer confusion and I think 2 choices -> 3 choices is gonna be a lot clearer than 2 choices -> 4 choices for users.

Reply:
Confirmed with Yi and Qinging that we'll resolve this before GA. We're also open to keep the name virtualMachines as the type.